tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/editor/BallerinaFoldingBuilder.java

```diff
@@ -0,0 +1,300 @@
+            @NotNull PsiElement root) {
+        // Get all documentation nodes.
+        Collection<BallerinaDocumentationString> docStrings = PsiTreeUtil.findChildrenOfType(root,
+                BallerinaDocumentationString.class);
+        for (BallerinaDocumentationString docString : docStrings) {
+            if (docString != null) {
+                // Calculate the start and end offsets.
+                int startOffset = docString.getTextRange().getStartOffset();
+                int endOffset = docString.getTextRange().getEndOffset();
+                // Add the new folding descriptor.
+                descriptors.add(new NamedFoldingDescriptor(docString, startOffset, endOffset, null, "# ..."));
+            }
+        }
+    }
+
+    private void buildAnnotationFoldingRegions(@NotNull List<FoldingDescriptor> descriptors, @NotNull PsiElement root) {
+        Collection<BallerinaAnnotationAttachment> annotations = PsiTreeUtil.findChildrenOfType(root,
+                BallerinaAnnotationAttachment.class);
+        for (BallerinaAnnotationAttachment annotation : annotations) {
+            // Get the annotation body. This is used to calculate the start offset.
+            BallerinaRecordLiteral annotationBody = PsiTreeUtil.getChildOfType(annotation,
+                    BallerinaRecordLiteral.class);
+            if (annotationBody == null) {
+                continue;
+            }
+            // Add folding descriptor.
+            addFoldingDescriptor(descriptors, annotation, annotationBody, false);
+        }
+    }
+
+    private void addFoldingDescriptor(@NotNull List<FoldingDescriptor> descriptors, PsiElement node,
+            PsiElement bodyNode, boolean includePrevious) {
+
+        PsiElement startNode = bodyNode;
+        if (includePrevious) {
+            PsiElement prevSibling = bodyNode.getPrevSibling();
+            // Sometimes the body node might start with a comment node.
+            while (prevSibling != null && (prevSibling instanceof PsiComment || prevSibling instanceof PsiWhiteSpace)) {
+                prevSibling = prevSibling.getPrevSibling();
+            }
+            startNode = prevSibling;
+        }
+
+        if (startNode != null) {
+            // Calculate the start and end offsets.
+            int startOffset = startNode.getTextRange().getStartOffset();
+            int endOffset = node.getTextRange().getEndOffset();
+            // Add the new folding descriptor.
+            descriptors.add(new NamedFoldingDescriptor(node, startOffset, endOffset, null, "{...}"));
+        }
+    }
+
+    private void buildMultiCommentFoldingRegions(@NotNull List<FoldingDescriptor> descriptors,
+            @NotNull PsiElement root) {
+
+        Collection<PsiComment> comments = PsiTreeUtil.findChildrenOfType(root, PsiComment.class);
+
+        for (PsiComment comment : comments) {
+            PsiElement prevSibling = getPreviousElement(comment);
+            if (!(prevSibling instanceof PsiComment)) {
```

Comment:
Invert the `if` logic here to remove the large nested block.

Reply:
Changed.